src/datamanager.go

```diff
@@ -0,0 +1,418 @@
+	logger.WithFields(logrus.Fields{
+		"operation": operation,
+		"namespace": namespace,
+		"leader":    ns.Leader,
+		"ns":        ns,
+	}).Info("ReadLeader successfully")
+
+	return ns.Leader, nil //returning copy
+}
+
+func (dm *DataManager) ReadAllLeaders() []LeaderController {
+	dm.mu.RLock()         // Read lock to allow multiple concurrent reads
+	defer dm.mu.RUnlock() // Ensure the lock is always released
+	operation := "ReadAllLeaders"
+
+	// Start the operation log
+	logger.WithFields(logrus.Fields{
+		"operation": operation,
+	}).Debug("Attempting to read namespace data")
+
+	var allLeaders []LeaderController
+
+	for _, data := range dm.namespaces {
+		allLeaders = append(allLeaders, data.Leader)
+	}
+	// Log success
+	logger.WithFields(logrus.Fields{
+		"operation": operation,
+		"allLeader": allLeaders,
+	}).Info("ReadAllLeaders successfully")
+
+	return allLeaders //returning copy
+}
+
+func (dm *DataManager) UpdateLeader(namespace string, leader LeaderController) error {
+	dm.mu.Lock()         // Read lock to allow multiple concurrent reads
+	defer dm.mu.Unlock() // Ensure the lock is always released
+	operation := "UpdateLeader"
+
+	// Ensure namespace exists
+	if _, exists := dm.namespaces[namespace]; !exists {
+		logger.WithFields(logrus.Fields{
+			"operation": operation,
+			"namespace": namespace,
+		}).Error("NamespaceData read failed")
+		err := fmt.Errorf("namespace '%s' not found", namespace)
+		return err
+	}
+
+	ns := dm.namespaces[namespace]
+	ns.Leader = leader
+	ns.Timestamp = time.Now() // Update timestamp on modification
+	dm.namespaces[namespace] = ns
+
+	logger.WithFields(logrus.Fields{
+		"operation": operation,
+		"namespace": namespace,
+		"leader":    ns.Leader,
+		"time":      ns.Timestamp,
+	}).Info("UpdateLeader data successfully")
+	return nil
+}
+
+func (dm *DataManager) ReadApp(namespace string) (map[string]App, error) {
+	dm.mu.RLock()         // Read lock to allow multiple concurrent reads
+	defer dm.mu.RUnlock() // Ensure the lock is always released
+	operation := "ReadLeader"
+
+	// Ensure namespace exists
+	if _, exists := dm.namespaces[namespace]; !exists {
+		logger.WithFields(logrus.Fields{
+			"operation": operation,
+			"namespace": namespace,
+		}).Error("NamespaceData read failed")
+		err := fmt.Errorf("namespace '%s' not found", namespace)
+		return map[string]App{}, err
+	}
+
+	ns := dm.namespaces[namespace]
+
+	// Log success
+	logger.WithFields(logrus.Fields{
+		"operation": operation,
+		"namespace": namespace,
+		"apps":      ns.Apps,
+	}).Info("ReadApp successfully")
+
+	return ns.Apps, nil //returning copy
+}
+
+func (dm *DataManager) ReadAllApps() map[string]App {
+	dm.mu.RLock()         // Read lock to allow multiple concurrent reads
+	defer dm.mu.RUnlock() // Ensure the lock is always released
+	operation := "ReadAllApps"
+
+	allApps := make(map[string]App)
+
+	for _, data := range dm.namespaces {
+		for appId, appData := range data.Apps {
+			allApps[appId] = appData
```

Comment:
If multiple namespaces contain same app (possible if multiple clusters have same deployment), this will overwrite

Reply:
Done